### PR TITLE
fix #16494

### LINKED ISSUE #16494: `1.0 / abs(-0.0)` returns -inf instead of inf
* abs(-0.0) returns -0.0 instead of 0.0, which causes further issues, eg:
`1.0 / abs(-0.0)` returns -inf instead of inf
* likewise, it should behave like fabs even for NaN (ie setting the sign bit) 

### Example1: -0.0
```nim
doAssert 1.0 / abs(-0.0) == Inf
```

### Current Output
fails

### Expected Output
works

### Example 2: same issue with nans
```nim
when true:
  proc fabs(a: cdouble): cdouble {.importc.}
  proc signbit(a: cdouble): cint {.importc, header: "<math.h>".}
  proc copySign(a, b: cdouble): cdouble {.importc: "copysign", header: "<math.h>".}
  let nan2 = copySign(NaN, -1.0)
  doAssert nan2.fabs.signbit == 0 # ok
  doAssert nan2.abs.signbit == 0 # fails
```

### Possible Solution
for implementation of `abs(SomeFloat)`, use importc fabs and fabsf in RT + VM (via vmops), with a fallback for js as usual; the fallback for js can use the same logic as for copySign(see https://github.com/nim-lang/Nim/pull/16406)

### Additional Information
devel 1.5.1 6d442a40a6f89572052d61aeb73ec26d1f3451ce

### PR DESCRIPTION
**Pull request by ringabout**
- [x] In the future PR, we could use `fabs` to refactor it.

### CODE REVIEW

**Comment by timotheecour** [comment]
> In the future PR, we could use fabs to refactor it.

yes please; fabs is 1.8X faster

```nim
when defined case2:
  proc abs1[T: float64 | float32](x: T): T {.noSideEffect, inline.} =
    if x < 0.0: -x
    elif x == 0.0: 0.0
    else: x

  proc fabs(x: cdouble): cdouble {.importc.}
  proc fabsf(x: cfloat): cfloat {.importc.}

  proc abs2[T: float64 | float32](x: T): T {.noSideEffect, inline.} =
    fabs(x)

  import times
  import random

  template work(algo) =
    let n = 1000_000_000
    let t = cpuTime()
    type T = float64
    var tot = T(0.0)
    for i in 0..<n:
      let x = cast[T](i*7) - tot
      tot += algo(x)
    let t2 = cpuTime()
    echo (astToStr(algo), t2-t, tot)

  proc main()=
    for i in 0..<3:
      work(abs1)
      work(abs2)
  main()
```

nim r -d:danger -d:case2 main
```
("abs1", 4.386831999999999, 3.458459517430266e-314)
("abs2", 2.328304999999999, 3.458459517430266e-314)
("abs1", 4.175053, 3.458459517430266e-314)
("abs2", 2.332252, 3.458459517430266e-314)
("abs1", 4.166519999999998, 3.458459517430266e-314)
("abs2", 2.271819000000001, 3.458459517430266e-314)
```

**Review by timotheecour — APPROVED** [review]

**Review by Clyybber — DISMISSED** [review]
The additional branching introduces unacceptable performance regressions.
Instead we should check the sign bit directly.

**Comment by xflywind** [comment]
Ok
fix the branch predicts

@Clyybber
> The additional branching introduces unacceptable performance regressions.
Instead we should check the sign bit directly.

```nim
proc abs1[T: float64 | float32](x: T): T {.noSideEffect, inline.} =
  if x < 0.0: -x
  elif x > 0.0: x
  elif x == 0.0: 0.0
  else: x
```

**Comment by xflywind** [comment]
For most of works:
```nim
proc abs1[T: float64 | float32](x: T): T {.noSideEffect, inline.} =
  if x < 0.0: -x
  elif x > 0.0: x
  elif x == 0.0: 0.0
  else: x

proc fabs(x: cdouble): cdouble {.importc.}
proc fabsf(x: cfloat): cfloat {.importc.}

proc abs2[T: float64 | float32](x: T): T {.noSideEffect, inline.} =
  fabs(x)

proc abs3[T: float64 | float32](x: T): T {.noSideEffect, inline.} =
  if x < 0.0: -x
  else: x

import times
import random

template work(algo) =
  let n = 1000_000_000
  let t = cpuTime()
  type T = float64
  var tot = T(0.0)
  for i in 0..<n:
    let x = cast[T](i*7) - tot
    tot += algo(x)
  let t2 = cpuTime()
  echo (astToStr(algo), t2-t, tot)

proc main()=
  for i in 0..<3:
    work(abs1)
    work(abs2)
    work(abs3)
main()
```

("abs1", 2.413, 3.458459517430266e-314)
("abs2", 2.742, 3.458459517430266e-314)
("abs3", 2.379, 3.458459517430266e-314)
("abs1", 2.456, 3.458459517430266e-314)
("abs2", 2.766, 3.458459517430266e-314)
("abs3", 2.450000000000001, 3.458459517430266e-314)
("abs1", 2.504000000000001, 3.458459517430266e-314)
("abs2", 2.811999999999998, 3.458459517430266e-314)
("abs3", 2.439, 3.458459517430266e-314)

**Comment by xflywind** [comment]
We haven't had `signbits`  in math yet. should `IEEE754_little_endian` or `IEEE754_big_endian` considered, we also to need consider that.

**Comment by xflywind** [comment]
```nim
proc abs1[T: float64 | float32](x: T): T {.noSideEffect, inline.} =
  if x < 0.0: -x
  elif x > 0.0: x
  elif x == 0.0: 0.0
  else: x

proc fabs(x: cdouble): cdouble {.importc.}
proc fabsf(x: cfloat): cfloat {.importc.}

proc abs2[T: float64 | float32](x: T): T {.noSideEffect, inline.} =
  fabs(x)

proc abs3[T: float64 | float32](x: T): T {.noSideEffect, inline.} =
  if x < 0.0: -x
  else: x

proc abs4[T: float64 | float32](x: T): T {.noSideEffect, inline.} =
  when T is float64:
    result = cast[float](cast[uint64](x) xor (1'u64 shr 63))
  else:
    result = cast[float](cast[uint32](x) xor (1'u32 shr 31))

import times
import random

template work(algo) =
  let n = 1000_000_000
  let t = cpuTime()
  type T = float64
  var tot = T(0.0)
  for i in 0..<n:
    let x = cast[T](i*7) - tot
    tot += algo(x)
  let t2 = cpuTime()
  echo (astToStr(algo), t2-t, tot)

proc main()=
  for i in 0..<3:
    work(abs1)
    work(abs2)
    work(abs3)
    work(abs4)
main()
```

**Comment by xflywind** [comment]
My approach doesn't have performance regression for most of float numbers except 0.0, -0.0 and NaN.

**Comment by Clyybber** [comment]
`abs4` in your example produces the best code, we should use that IMO and fall back to `abs1` on js.

**Comment by ringabout** [comment]
I can't find a proper bit operation implementation.

**Comment by ringabout** [comment]
Current implementation is not that bad

```nim
proc abs1[T: float64 | float32](x: T): T {.noSideEffect, inline.} =
  if x < 0.0: -x
  elif x > 0.0: x
  elif x == 0.0: 0.0
  else: x

proc fabs(x: cdouble): cdouble {.importc.}
proc fabsf(x: cfloat): cfloat {.importc.}

proc abs2[T: float64 | float32](x: T): T {.noSideEffect, inline.} =
  fabs(x)

proc abs3[T: float64 | float32](x: T): T {.noSideEffect, inline.} =
  if x < 0.0: -x
  else: x

import times
import random

template work(algo) =
  let n = 1000_000_000
  let t = cpuTime()
  type T = float64
  var tot = T(0.0)
  for i in 0..<n:
    let x = T(i*7) - tot
    tot += algo(x)
    tot += algo(-x)
  let t2 = cpuTime()
  echo (astToStr(algo), t2-t, tot)

proc main()=
  for i in 0..<3:
    work(abs1)
    work(abs2)
    work(abs3)
main()
```
-d:danger

**Comment by ringabout** [comment]
This PR just fix `abs(-0.0) == -0.0`.
- [x] In the next PR, we could use `fabs` in C backend and `abs` in JS backend.

File: lib/system.nim [hunk 1]
```diff
@@ -1522,10 +1522,12 @@ include "system/iterators_1"
 
 {.push stackTrace: off.}
 
-proc abs*(x: float64): float64 {.noSideEffect, inline.} =
-  if x < 0.0: -x else: x
-proc abs*(x: float32): float32 {.noSideEffect, inline.} =
-  if x < 0.0: -x else: x
+proc abs*[T: float64 | float32](x: T): T {.noSideEffect, inline.} =
+  if x < 0.0: -x
+  elif x > 0.0: x
```
Comment (Araq) [inline lib/system.nim]: The ``elif x > 0.0: x`` case can be merged with the final ``else: x``. I think. If not, add a comment please.

Comment (ringabout) [reply lib/system.nim]: Yeah, it can't handle `NaN`, I will add a comment.

Comment (Clyybber) [reply lib/system.nim]: This is what this PR did before and it introduced a heavy performance regression.
This extra branching seemed to help the predictor somewhat, but this benchmark shows that we are really just playing games with the branch predictor:
```nim
proc abs1(x: float64): float64 {.noSideEffect, noinline.} =
  if x < 0.0: -x
  elif x > 0.0: x
  elif x == 0.0: 0.0
  else: x

proc abs2(x: float64): float64 {.noSideEffect, noinline.} =
  if x < 0.0: -x
  elif x == 0.0: 0.0
  else: x

proc fabs(x: cdouble): cdouble {.importc.}

proc abs3(x: float64): float64 {.noSideEffect, noinline.} =
  fabs(x)

proc abs4(x: float64): float64 {.noSideEffect, noinline.} =
  if (cast[uint64](x) and 1'u64 shl 63) != 0: -x
  else: x

proc abs5(x: float64): float64 {.noSideEffect, noinline.} =
  cast[float64](cast[uint64](x) and not(1'u64 shl 63))

import bitops
proc abs6[T: float64](x: T): T {.noSideEffect, inline.} =
  var x = cast[uint64](x)
  clearBit(x, 1)
  result = cast[float64](x)

import times
import random

template work(algo) =
  let n = 1000_000_000
  let t = cpuTime()
  type T = float64
  var tot = T(0.0)
  for i in 0..<n:
    let x = T(i*7) - tot
    tot += algo(x)
    tot += algo(-x)
  let t2 = cpuTime()
  echo (astToStr(algo), t2-t, tot)

proc main()=
  work(abs1)
  work(abs2)
  work(abs3)
  work(abs4)
  work(abs5)
main()
```
depending on how big you set `n`, the branch versions start to win or loose against the bit twiddling version or `fabs` depending on your compiler/arch.

It's probably for the best if we just delegate to `fabs` which uses compiler builtins.

**Comment by ringabout** [comment]
Another PR of mine is blocked by this though

Ref https://github.com/nim-lang/Nim/issues/9125
But I'm not sure whether I should unify the C and JS backend regarding `round`.
```nim
import math

# proc abs(x: float): float {.importc: "fabs", header: "<math.h>".}

proc abs2*[T: float64 | float32](x: T): T {.noSideEffect, inline.} =
  if x < 0.0: -x
  elif x > 0.0: x
  elif x == 0.0: 0.0 # handle 0.0, -0.0
  else: x # handle NaN

proc copysign(x, y: float): float =
  if y > 0.0 or (y == 0.0 and 1.0 / y > 0.0):
    result = abs2(x)
  else:
    result = -abs2(x)

func round2*(x: float): float =
  let absx = abs2(x)
  var y = floor(absx)
  if (absx - y) >= 0.5:
    y += 1.0
  result = copySign(y, x)

proc hello() =
  echo round2(-0.0)
  echo round2(-1.5)
  echo round2(-2.5)

  echo round2(1.5)
  echo round2(2.5)

static: hello()
hello()
```

This implementation have the same results in C, JS and VM backend
```
-0.0
-2.0
-3.0
2.0
3.0
```
Round to sgn(x) * inf

BTW original windowsCC89 implementation is not correct.
```nim
  when windowsCC89:
    func round*[T: float32|float64](x: T): T =
      ## Windows compilers prior to MSVC 2012 do not implement 'round',
      ## 'roundl' or 'roundf'.
      result = if x < 0.0: ceil(x - T(0.5)) else: floor(x + T(0.5))
```
see https://github.com/nim-lang/Nim/issues/16434#issuecomment-749337617

**Comment by xflywind** [comment]
I want to reduce one call in JS backend, but it doesn't work
```nim
when defined(js):
  proc abs*[T: float64 | float32](x: T): T {.importc: "Math.abs".}
else:
  proc fabs(x: cdouble): cdouble {.importc: "fabs", header: "math.h".}
  proc fabsf(x: cfloat): cfloat {.importc: "fabsf", header: "math.h".}

  proc abs*[T: float64 | float32](x: T): T {.noSideEffect, inline.} =
    when nimvm:
      # TODO register VM ops
      if x < 0.0: result = -x
      elif x > 0.0: result = x
      elif x == 0.0: result = 0.0 # handle 0.0, -0.0
      else: result = x # handle NaN
    else:
      when T is float64:
        result = fabs(x)
      else:
        result = fabsf(x)
```

File: lib/system.nim [hunk 2]
```diff
@@ -1522,10 +1522,27 @@ include "system/iterators_1"
...
+    # TODO register VM ops
+    if x < 0.0: result = -x
+    elif x > 0.0: result = x
+    elif x == 0.0: result = 0.0 # handle 0.0, -0.0
+    else: result = x # handle NaN
+  else:
+    when T is float64:
+      result = fabs(x)
+    else:
+      result = fabsf(x)
```
Comment (timotheecour) [inline lib/system.nim]:
```suggestion
when defined(js):
  proc js_abs[T: SomeNumber](x: T): T {.importc: "Math.abs".}
else:
  proc c_fabs(x: cdouble): cdouble {.importc: "fabs", header: "math.h".}
  proc c_fabsf(x: cfloat): cfloat {.importc: "fabsf", header: "math.h".}

proc abs*[T: float64 | float32](x: T): T {.noSideEffect, inline.} =
  when nimvm:
    if x < 0.0: result = -x
    elif x == 0.0: result = 0.0 # handle 0.0, -0.0
    else: result = x # handle NaN, > 0
  else:
    when defined(js): result = js_abs(x)
    else:
      when T is float64:
        result = c_fabs(x)
      else:
        result = c_fabsf(x)
```

made a few changes:
* 1 less branch
* no need for vmops here, I feel like in this case, it's fast enough in VM
* changed fabs (etc) to c_fabs
* made js code more explicit (avoiding conflation with c)

File: lib/system.nim [hunk 3]
```diff
@@ -1522,10 +1522,26 @@ include "system/iterators_1"
...
-proc abs*(x: float64): float64 {.noSideEffect, inline.} =
-  if x < 0.0: -x else: x
-proc abs*(x: float32): float32 {.noSideEffect, inline.} =
-  if x < 0.0: -x else: x
+
+when defined(js):
+  proc js_abs[T: SomeNumber](x: T): T {.importc: "Math.abs".}
+else:
+  proc c_fabs(x: cdouble): cdouble {.importc: "fabs", header: "math.h".}
+  proc c_fabsf(x: cfloat): cfloat {.importc: "fabsf", header: "math.h".}
```
Comment (timotheecour) [inline lib/system.nim]:
```suggestion
  proc c_fabs(x: cdouble): cdouble {.importc: "fabs", header: "<math.h>".}
  proc c_fabsf(x: cfloat): cfloat {.importc: "fabsf", header: "<math.h>".}
```

File: tests/stdlib/tmath_misc.nim [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+import math
+
+proc main() =
+  block:
+    doAssert 1.0 / abs(-0.0) == Inf
+    doAssert 1.0 / abs(0.0) == Inf
+    doAssert -1.0 / abs(-0.0) == -Inf
+    doAssert -1.0 / abs(0.0) == -Inf
+
```
Comment (timotheecour) [inline tests/stdlib/tmath_misc.nim:17]:
```suggestion
doAssert abs(0.0) == 0.0
doAssert abs(0.0'f32) == 0.0'f32
```

File: tests/stdlib/tmath_misc.nim [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+import math
+
+proc main() =
+  block:
+    doAssert 1.0 / abs(-0.0) == Inf
+    doAssert 1.0 / abs(0.0) == Inf
+    doAssert -1.0 / abs(-0.0) == -Inf
+    doAssert -1.0 / abs(0.0) == -Inf
+doAssert abs(0.0) == 0.0
+doAssert abs(0.0'f32) == 0.0'f32
```
Comment (timotheecour) [inline tests/stdlib/tmath_misc.nim]: (my bad...)

```suggestion
    doAssert abs(0.0) == 0.0
    doAssert abs(0.0'f32) == 0.0'f32

```

**Review by timotheecour — APPROVED** [review]
LGTM
future work:
`make abs(someNaN)` correct in VM (it should set the sign bit, as done in RT) by using a vmops; this is a niche case (and pre-existing) so doesn't block this PR

CI failure unrelated: https://github.com/nim-lang/Nim/issues/16523

**Pull request merged**